### PR TITLE
[Inductor] Extend Pattern Matcher to Match Equivalent Function Invocation

### LINKED ISSUE #104391: Extend Pattern Matcher to Match Equivalent Function Invocations
### 🐛 Describe the bug

If you add a print statement to `should_pad_addmm` in `pad_mm`.py :

```
--- a/torch/_inductor/fx_passes/pad_mm.py
+++ b/torch/_inductor/fx_passes/pad_mm.py
@@ -82,6 +82,7 @@ def addmm_pattern(input, mat1, mat2, beta, alpha):
 
 
 def should_pad_addmm(match):
+    print("TESTING FOR REPLACEMENT")
     mat1, mat2, input = fetch_fake_tensors(match, ("mat1", "mat2", "input"))
     return should_pad_common(mat1, mat2, input) and should_pad_bench(
         mat1, mat2, torch.ops.aten.addmm, input=input 
```

and then run the following script:

```
import torch

@torch.compile()
def test_mm_padding(inp, x, y):
    return torch.ops.aten.addmm(inp, x, y)

@torch.compile()
def test_mm_padding2(inp, x, y):
    return torch.ops.aten.addmm(inp, x, y, alpha=1.0, beta=1.0)

inps = [torch.rand([20]), torch.rand([20, 20]), torch.rand([20, 20])]
inps = [x.to(device="cuda") for x in inps]

print("Running first")
test_mm_padding(*inps)

print("Running second")
test_mm_padding2(*inps)
```

You'll see that only second addmm invocation which explicitly passes in alpha and beta get matched, even though both addmm invocations are equivalent. That is because we are matching on exact [# of args and kwargs](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py#L343-L347). 

We should extend the pattern matcher so that equivalent invocations get matched. A relevant function here is [normalize_function](https://github.com/pytorch/pytorch/blob/732067e5c30519df5c8906913825bd905e9185fa/torch/fx/operator_schemas.py#L247). 

A couple things you may run into:

- there's an annoying thing where pytorch replaces "self" with "input
- this might break existing replacement patterns that are directly introspecting the fx node. replacement patterns should mostly be fine but graph patterns are more likely to break. we could have an option to the patterns that would canonicalize the calling convention to the convention that its been specified as. There aren't tooooo many existing patterns so it shouldn't be terrible to update them.

shout out @aakhundov for spotting this !

### Versions

master

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #104391

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### CODE REVIEW

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -342,17 +342,29 @@ def __repr__(self):
...
+        if not self._match_fns(node):
             return FailedMatch(f"function_mismatch: node={node}, pattern={self}")
 
         if not self._match_users(node, ctx):
             return FailedMatch(f"multiple_users {node}")
 
-        node_items, node_spec = self.flatten(node.args, node.kwargs)
+        _args = node.args
+        _kwargs = node.kwargs
+        if len(_args) != len(self.args) or len(_kwargs) != len(self.kwargs):
```
Comment (aakhundov) [inline torch/_inductor/pattern_matcher.py:353]: Should we make sure that `self.args` and `self.kwargs` are also normalized?

Comment (yanboliang) [reply torch/_inductor/pattern_matcher.py:353]: I was thinking about this, but ```self.args``` and ```self.kwargs``` are from the the pattern user defined, where we can ensure they are normalized. And I'm not sure if we allow users intend to write unnormalized pattern in some scenarios.

Comment (aakhundov) [reply torch/_inductor/pattern_matcher.py:353]: [Here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/freezing_patterns.py#L153-L155), for example, `alpha` and `beta` are not explicitly specified. I guess, the pattern author meant to use the defaults, but then we have an unnormalized pattern?

**Comment by eellison** [comment]
will get to this tomorrow morning.

File: torch/_inductor/pattern_matcher.py [hunk 2]
```diff
@@ -342,17 +342,29 @@ def __repr__(self):
...
+            from torch.fx.operator_schemas import normalize_function
+
+            arg_types = [type(node.args[i]) for i in range(len(node.args))]
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:356]: i'm not sure how this is working, since `arg_types` here will give you `torch.fx.node.Node`

File: test/inductor/test_pattern_matcher.py [hunk 3]
```diff
@@ -664,6 +665,37 @@ def fn2(inp, a, b):
         _, (code) = run_and_get_code(fn2, args[0], args[1], args[2])
         FileCheck().check_not("extern_kernels.addmm(").run(code[0])
 
+    def test_match_equivalent_function_invocations(self):
+        # pad_mm.py only registered `addmm_pattern` which explicitly passes in alpha and beta,
+        # since they have default values, we should also match the equivalent invocations like
+        # the following two cases.
+        @torch.compile()
```
Comment (eellison) [inline test/inductor/test_pattern_matcher.py:672]: Can we also add a test in the other direction, say where the user registers `addmm` without alpha and beta, but alpha=1. and beta=1. are passed in? 

And where the user registers a pattern with just alpha handled but not beta ?

Note: torch_dispatch will remove default args, so we might need to register passes that would add/remove them from node to exercise our logic

File: torch/_inductor/fx_passes/split_cat.py [hunk 4]
```diff
@@ -781,7 +781,7 @@ def get_transform_params(
         split_dim = unbind_node.kwargs["dim"]
         transform_params_list = []
         for user_node, user_inputs in zip(next_users, user_inputs_list):
-            cat_dim = get_arg_value(user_node, 1, "dim")
+            cat_dim = get_arg_value(user_node, 1, "dim") or 0
```
Comment (eellison) [inline torch/_inductor/fx_passes/split_cat.py:784]: why was this necessary?

File: test/inductor/test_pattern_matcher.py [hunk 5]
```diff
@@ -664,6 +665,37 @@ def fn2(inp, a, b):
...
+        def f1(inp, x, y):
+            return torch.ops.aten.addmm(inp, x, y)
+
+        @torch.compile()
+        def f2(inp, x, y):
+            return torch.ops.aten.addmm(inp, x, y, alpha=1, beta=1)
```
Comment (eellison) [inline test/inductor/test_pattern_matcher.py:678]: These are actually the same test case, since torch_dispatch will remove default args. We might need to add custom passes to exercise normalization logic.

**Review by eellison — COMMENTED** [review]
looking good, but I think we need a few changes to more exhaustively test this

File: test/inductor/test_pattern_matcher.py [hunk 6]
```diff
@@ -664,6 +665,37 @@ def fn2(inp, a, b):
...
+            with unittest.mock.patch.dict(
+                os.environ, {"TORCHINDUCTOR_PATTERN_MATCH_DEBUG": "addmm"}
+            ):
+                with self.assertLogs(
+                    logger="torch._inductor.pattern_matcher", level="WARNING"
+                ) as logs:
+                    fn(*inps)
+                    self.assertEqual(
+                        sum(
+                            msg.find("addmm(arg0_1, arg1_1, arg2_1) Match(...") != -1
```
[the remark this answers is not in the document]
Comment (yanboliang) [reply test/inductor/test_pattern_matcher.py:693]: Agreed, I think registering a testing purpose graph transformation is a good idea.

File: torch/_inductor/fx_passes/split_cat.py (hunk 4 above)
Comment (yanboliang) [reply torch/_inductor/fx_passes/split_cat.py:784]: This fixed an existing bug, not relevant to this change directly.

File: test/inductor/test_pattern_matcher.py [hunk 7]
```diff
@@ -763,6 +765,154 @@ def fn2(inp, a, b):
...
+        )
+        def addmm_replacement(match: Match, inp, mat1, mat2, beta):
+            nonlocal counter
+            counter += 1
+
+            def repl(inp, x1, x2):
+                return x1 @ x2 + inp
+
+            with V.fake_mode:
+                match.replace_by_example(repl, [inp, mat1, mat2])
```
Comment (eellison) [inline test/inductor/test_pattern_matcher.py:903]: It would be great if this test case could work across all three invocations as part of this PR, since these three are all equivalent.  We would want to do something where we update the an arg to the fx node to drop specified inputs when they are equal to the default value and that arg is present in the pattern it is matching.

**Review by eellison — COMMENTED** [review]
Thanks for extending the tests - would be great to have it always working for equivalent invocations

**Review by jansel — APPROVED** [review]

File: test/inductor/test_pattern_matcher.py [hunk 8]
```diff
@@ -775,6 +777,154 @@ def fn2(inp, a, b):
...
+
+        with unittest.mock.patch(
+            "torch._inductor.fx_passes.post_grad.pass_patterns",
+            torch._inductor.fx_passes.post_grad.pass_patterns + [test_pass],
+        ):
+            for fn in (f0, f1, f2):
+                counter = 0
+                expected = fn(*copy.deepcopy(args))
+                actual = torch.compile(fn)(*copy.deepcopy(args))
+                self.assertEqual(counter, 1)
```
Comment (eellison) [inline test/inductor/test_pattern_matcher.py:878]: how does this work, where does the counter get updated ?

**Comment by jansel** [comment]
@yanboliang can you resubmit this?

**Comment by yanboliang** [comment]
> @yanboliang can you resubmit this?

Will do it, still working on debugging the internal failures.

**Comment by eellison** [comment]
@yanboliang please re-request when you want another review

**Comment by eellison** [comment]
how's this going ?
